### PR TITLE
Early stopping, learning rate and noise decay

### LINKED ISSUE #398: Early Stopping, Learning rate and noise decay
Implement the best practices from multi-agent Rl community and stablebaselines3 into our algorithm. Further analyse similarities between petting zoo multi-agent implementation to current RL implementation of Assume. (https://towardsdatascience.com/multi-agent-deep-reinforcement-learning-in-15-lines-of-code-using-pettingzoo-e0b963c0820b)

### PR DESCRIPTION
**Pull request by mthede**
# Pull Request

## Related Issue
Closes #398 

## Description
Implement best practices from multi-agent RL community and stable-baselines3 into our algorithm

## Changes Proposed
- No early stopping by default
- Linear schedule for learning rate and action noise decay 
- Basic scheduling options adapted from SB3

## Testing
[Describe the testing you've done, including any specific test cases or scenarios]

## Checklist
Please check all applicable items:

- [x] Code changes are sufficiently documented (docstrings, inline comments, `doc` folder updates)
- [x] New unit tests added for new features or bug fixes
- [x] Existing tests pass with the changes
- [ ] Reinforcement learning examples are operational (for DRL-related changes)
- [x] Code tested with both local and Docker databases
- [x] Code follows project style guidelines and best practices
- [x] Changes are backwards compatible, or deprecation notices added
- [x] New dependencies added to `pyproject.toml`
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included
- [x] Consent to release this PR's code under the GNU Affero General Public License v3.0

## Additional Notes (if applicable)
- Scheduling option only available in codebase, no config item provided yet
- The structure (learning role, learning algorithm, learning strategy) made it sometimes difficult to "get the necessary information" to the right place
- We've therefore decided to update the action noise with every policy update instead of every time the actions are sampled
- The NormalActionNoise was adjusted because the previous implementation enabled only exponential decay, now the dt value can be scheduled arbitrarily

### CODE REVIEW

File: assume/world.py [hunk 1]
```diff
@@ -479,6 +480,14 @@ def create_unit(
...
         )
 
+    def _add_simulation_horizon_to_learning_role(self):
+        """
+        Add simulation horizon to learning role for calculation of decay.
+        """
+
+        self.learning_role.start = datetime2timestamp(self.start)
+        self.learning_role.end = datetime2timestamp(self.end)
+
```
Comment (kim-mskw) [inline assume/world.py:490]: Why did you decide to make this a function of world and not the learning_role?

**Comment by kim-mskw** [comment]
Open discussion: Do we want to allow people to activating and deactivate the scheduling in the learning config?

**Comment by kim-mskw** [comment]
> Open discussion: Do we want to allow people to activate and deactivate the scheduling in the learning config?
From the discussion in the development meeting we want to have it in the config but keep it as optional. If it is not defined, it is not used.

File: assume/world.py (hunk 1 above)
Comment (mthede) [reply assume/world.py:490]: I think I just tried to keep the structure of the function below __add_bidding_strategies_to_learning_role_ 
But I see, it would probably be better as a setter function of the LearningRole with start, end passed as args.

File: docs/source/learning_algorithm.rst [hunk 2]
```diff
@@ -43,13 +43,14 @@ The following table shows the options that can be adjusted and gives a short exp
...
-  noise_dt                                Determines how quickly the noise weakens over time.
+  noise_sigma                             The standard deviation of the distribution used to draw the noise, which is added to the actions and forces exploration.
+  noise_dt                                Determines how quickly the noise weakens over time / used for noise scheduling.
   noise_scale                             The scale of the noise, which is multiplied by the noise drawn from the distribution.
   early_stopping_steps                    The number of steps considered for early stopping. If the moving average reward does not improve over this number of steps, the learning is stopped.
   early_stopping_threshold                The value by which the average reward needs to improve to avoid early stopping.
  ======================================== ==========================================================================================================
 
+Options for (linear) scheduling of learning rate and action noise are available, but not settable in the config file.
```
Comment (maurerle) [inline docs/source/learning_algorithm.rst:53]: This is not true anymore?
So the two parameters should be added to the table above right?

**Review by maurerle — APPROVED** [review]

File: docs/source/learning_algorithm.rst (hunk 2 above)
Comment (mthede) [reply docs/source/learning_algorithm.rst:53]: Yeah, you're right. Missed that - thanks! I'm adding them now.

File: assume/reinforcement_learning/learning_utils.py [hunk 3]
```diff
@@ -64,10 +69,18 @@ def __init__(self, action_dimension, mu=0.0, sigma=0.1, scale=1.0, dt=0.9998):
         self.dt = dt
 
     def noise(self):
-        noise = self.scale * np.random.normal(self.mu, self.sigma, self.act_dimension)
-        self.scale = self.dt * self.scale  # if self.scale >= 0.1 else self.scale
+        # TODO: document changes to normal action noise and different usage of dt parameter?
```
Comment (mthede) [inline assume/reinforcement_learning/learning_utils.py:72]: I've changed the usage of noise_dt. Before, it was only possible to use it for exponential decay. Now, it can be set by the scheduling function independently (although currently only linear schedule implemented).

Where do I need to document those changes? And can I delete line 78, if it's documented and approved by everyone?

**Comment by kim-mskw** [comment]
@nick-harder, this PR is ready and already approved by @maurerle. From a DRL perspective, I'd like your "Go" before merging anyway.

**Review by maurerle — COMMENTED** [review]
There are some merge conflicts to be solved as well, right? :)

File: docs/source/release_notes.rst [hunk 4]
```diff
@@ -13,6 +13,10 @@ Upcoming Release
   The features in this section are not released yet, but will be part of the next release! To use the features already you have to install the main branch,
   e.g. ``pip install git+https://github.com/assume-framework/assume``
 
+**New Features:**
```
Comment (nick-harder) [inline docs/source/release_notes.rst:16]: would be good to also mention the default behavior that if not specified they are not active

File: assume/reinforcement_learning/learning_role.py [hunk 5]
```diff
@@ -70,6 +84,20 @@ def __init__(
...
+            # constant schedule as no config item stating it should be scheduled is present
+            self.lr_schedule = get_schedule_fn(self.learning_rate)
+
+        noise_dt = learning_config.get("noise_dt", 1)
+        use_noise_schedule = learning_config.get("use_noise_schedule", False)
+        if use_noise_schedule:
+            self.noise_schedule = get_schedule_fn(linear_schedule(noise_dt))
+        else:
+            # constant schedule as no config item stating it should be scheduled is present
+            self.noise_schedule = get_schedule_fn(noise_dt)
```
[the remark this answers is not in the document]
Comment (nick-harder) [reply assume/reinforcement_learning/learning_role.py:100]: I would agree that having this as a name or a dict like {"linear": 5}. would be better also for the future

File: assume/reinforcement_learning/learning_role.py [hunk 6]
```diff
@@ -329,6 +367,9 @@ def compare_and_save_policies(self, metrics: dict) -> bool:
                         logger.info(
                             f"Stopping training as no improvement above {self.early_stopping_threshold} in last {self.early_stopping_steps} evaluations for {metric}"
                         )
+                        logger.info(
```
Comment (nick-harder) [inline assume/reinforcement_learning/learning_role.py:370]: maybe add the check if self.decay was true and display this info that there were these scheduelrs and maybe not the optimum was reached?

File: examples/inputs/example_02a/config.yaml
[the remark this answers is not in the document]
Comment (mthede) [reply examples/inputs/example_02a/config.yaml:1]: that was by mistake - sorry!

File: assume/reinforcement_learning/learning_role.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mthede) [reply assume/reinforcement_learning/learning_role.py]: thanks for the valueable feedback, I implemented your suggested solution!

File: examples/notebooks/11_redispatch.ipynb
[the remark this answers is not in the document]
Comment (mthede) [reply examples/notebooks/11_redispatch.ipynb:1]: I can't really explain that... I didn't change anything in that notebook

[the remark this answers is not in the document]
Comment (mthede) [reply examples/notebooks/11_redispatch.ipynb:1]: removed the changes
